pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,290 @@
+func (f *ftpFileSystem) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		f.logger = l
+	}
+}
+
+// UseMetrics sets the metrics for the ftpFileSystem client which asserts the Metrics interface.
+// Currently not implemented.
+func (*ftpFileSystem) UseMetrics(_ interface{}) {
+
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *ftpFileSystem) Connect() {
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	const dialTimeout = 5 * time.Second
+
+	conn, err := ftp.Dial(ftpServer, ftp.DialWithTimeout(dialTimeout))
+	if err != nil {
+		f.logger.Errorf("Connection Failed Error: %v", err)
+		return
+	}
+
+	f.conn = &Conn{conn}
+
+	f.logger.Logf("Connection Success Connected to FTP Server : %v", ftpServer)
+
+	err = conn.Login(f.config.User, f.config.Password)
+	if err != nil {
+		f.logger.Errorf("Login Failed Error: %v", err)
+		return
+	}
+
+	f.logger.Logf("Login Success Current remote location: %q", f.config.RemoteDir)
+}
+
+// Create creates an empty file on the FTP server.
+func (f *ftpFileSystem) Create(name string) (File, error) {
+	if name == "" {
+		f.logger.Errorf("Create_File Failed Provide a valid filename: %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	emptyReader := new(bytes.Buffer)
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.Stor(filePath, emptyReader)
+	if err != nil {
+		f.logger.Errorf("Create_File Failed Error creating file  with path %q : %v", filePath, err)
+		return nil, err
+	}
+
+	s := strings.Split(filePath, "/")
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Create_File Failed Error : %v", err)
+		return nil, err
+	}
+
+	f.logger.Logf("Create_File Success Successfully created file %s at %q", name, filePath)
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s[len(s)-1],
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server. Here, os.FileMode is unused.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	if name == "" {
+		f.logger.Errorf("Mkdir Failed Provide a valid directory : %v", errEmptyDirectory)
+		return errEmptyDirectory
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.MakeDir(filePath)
+	if err != nil {
+		f.logger.Errorf("Mkdir Failed Error creating directory at %q : %v", filePath, err)
+		return err
+	}
+
+	f.logger.Logf("Mkdir Success %s successfully created", name)
+
+	return nil
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+// The directories are not created if even one directory exist.
+func (f *ftpFileSystem) MkdirAll(path string, _ os.FileMode) error {
+	if path == "" {
+		f.logger.Errorf("MkdirAll Failed Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+
+	dirs := strings.Split(path, "/")
```

Comment:
filepath.Split

Reply:
Same, `path.Split`

Reply:
path.Split returns the file and directory ..... here this function creates all the directories iteratively .....so I do need all the directories to check whether they exist and if they don't .. then to create them .... I don't think I can use path.Split. If  I use path.Split I would have to put it inside the iteration to get the directory each time .

Reply:
https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/os/path.go;l=18 read up on how the built in `MkdirAll` is done, it starts with the equivalent of a `path.Split`